machida/wallaroo.py

```diff
@@ -105,24 +106,118 @@ def build(self):
         return self._actions
 
 
+def computation(name):
+    def wrapped(computation_function):
+        @wraps(computation_function)
+        class C:
+            def name(self):
+                return name
+            def compute(self, data):
+                return computation_function(data)
+
+        return C
+
+    return wrapped
+
+
+def state_computation(name):
+    def wrapped(computation_function):
+        @wraps(computation_function)
+        class C:
+            def name(self):
+                return name
+            def compute(self, data, state):
+                return computation_function(data, state)
+
+        return C()
+
+    return wrapped
+
+
+def computation_multi(name):
+    def wrapped(computation_function):
+        @wraps(computation_function)
+        class C:
+            def name(self):
+                return name
+            def compute_multi(self, data):
+                return computation_function(data)
+
+        return C
+
+    return wrapped
+
+
+def state_computation_multi(name):
+    def wrapped(computation_function):
+        @wraps(computation_function)
+        class C:
+            def name(self):
+                return name
+            def compute_multi(self, data, state):
+                return computation_function(data, state)
+
+        return C
+
+    return wrapped
+
+
+def state(clz):
+    clz.build = classmethod(lambda(c): clz())
+    return clz
+
+
+def partition(fn):
+    @wraps(fn)
+    class C:
+        def partition(self, data):
+            return fn(data)
+    return C()
+
+
+def decoder(header_length, length_fmt):
+    def wrapped(decoder_function):
+        @wraps(decoder_function)
+        class C:
+            def header_length(self):
+                return header_length
+            def payload_length(self, bs):
+                return struct.unpack(length_fmt, bs)[0]
+            def decode(self, bs):
+                return decoder_function(bs)
+
+        return C()
+
+    return wrapped
+
+
+def encoder(encoder_function):
+    class C:
+        @wraps(encoder_function)
+        def encode(self, data):
+            return encoder_function(data)
+
+    return C()
+
+
 class TCPSourceConfig(object):
-    def __init__(self, host, port, decoder):
```

Comment:
Somehow `decoder`s and `encoder`s got switch around here and in `TCPSinkConfig`.